### ISSUE TITLE
`prefer-const` is reported when assignment occurs in different scope.

### ISSUE THREAD

**Issue by strangedev** [eslint/eslint#14977]
**Tell us about your environment**

Environment Info:

Node version: v16.7.0
npm version: v7.20.6
Local ESLint version: Not found
Global ESLint version: Not found
Operating System: linux 5.13.10-arch1-1

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

We're using the `@typescript-eslint/parser`.

**Please show your full configuration:**

https://github.com/thenativeweb/eslint-config-es/blob/main/browser.js which extends
https://github.com/thenativeweb/eslint-config-es/blob/main/node.js

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

We're shipping the new eslint version in our build tool [roboter](https://github.com/thenativeweb/roboter). After updating, we've encountered a couple of reported issues that hadn't been reported before while linting one of our packages. We believe the `prefer-const` rule is reported falsely in the following case. Eslint is called by roboter. It uses the JS API [as seen here](https://github.com/thenativeweb/roboter/blob/main/lib/steps/analyze/lintCode.ts).

```ts
const getBrokenUrls = async function ([...]): Promise<BrokenUrl[]> {
  
  [...]

  // eslint-disable-next-line no-constant-condition, @typescript-eslint/no-unnecessary-condition
  while (true) {
    
    [...]

    let page: Page;

    try {
      page = await getPage({ url: currentUrl });
    } catch {
      brokenUrls.push({ url: currentUrl, referrers: referrers[currentUrl] });
      onBrokenUrl({ url: currentUrl, referrers: referrers[currentUrl] });
      continue;
    }

    if (!currentUrl.startsWith(baseUrl)) {
      continue;
    }

    const externalUrls = getExternalUrlsFromPage({ page, baseUrl });

   [...]
};
```

**What did you expect to happen?**

The `prefer-const` rule should not apply to the file shown, as `page` is declared in a different scope than it is first assigned.
The declaration of `page` can't be moved into the `try` block without moving other statements into the block as well.

**What actually happened? Please copy-paste the actual, raw output from ESLint.**

Eslint reports the error: `error  'page' is never reassigned. Use 'const' instead  prefer-const`.

**Steps to reproduce this issue:**

1. `git clone git@github.com:thenativeweb/thenativeweb-ux.git --branch dependabot/npm_and_yarn/roboter-12.0.0`
2. `cd thenativeweb-ux && npm install`
3. `npx roboter`

**Are you willing to submit a pull request to fix this bug?**

Yes

**Comment by mdjermanovic** [comment]
Hi @strangedev, thanks for the issue!

You're right, `page` in your example shouldn't be reported by `prefer-const`, but I can't reproduce this in ESLint v7.32.0 with the default parser.

 Here's a [demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHByZWZlci1jb25zdDogWyAnZXJyb3InLCB7XG4gICAgZGVzdHJ1Y3R1cmluZzogJ2FueScsXG4gICAgaWdub3JlUmVhZEJlZm9yZUFzc2lnbjogdHJ1ZVxuICB9XSAqL1xuXG5jb25zdCBnZXRCcm9rZW5VcmxzID0gYXN5bmMgZnVuY3Rpb24gKCkge1xuICBcbiAgd2hpbGUgKHRydWUpIHtcblxuICAgIGxldCBwYWdlO1xuXG4gICAgdHJ5IHtcbiAgICAgIHBhZ2UgPSBhd2FpdCBnZXRQYWdlKHsgdXJsOiBjdXJyZW50VXJsIH0pO1xuICAgIH0gY2F0Y2gge1xuICAgICAgYnJva2VuVXJscy5wdXNoKHsgdXJsOiBjdXJyZW50VXJsLCByZWZlcnJlcnM6IHJlZmVycmVyc1tjdXJyZW50VXJsXSB9KTtcbiAgICAgIG9uQnJva2VuVXJsKHsgdXJsOiBjdXJyZW50VXJsLCByZWZlcnJlcnM6IHJlZmVycmVyc1tjdXJyZW50VXJsXSB9KTtcbiAgICAgIGNvbnRpbnVlO1xuICAgIH1cblxuICAgIGlmICghY3VycmVudFVybC5zdGFydHNXaXRoKGJhc2VVcmwpKSB7XG4gICAgICBjb250aW51ZTtcbiAgICB9XG5cbiAgICBjb25zdCBleHRlcm5hbFVybHMgPSBnZXRFeHRlcm5hbFVybHNGcm9tUGFnZSh7IHBhZ2UsIGJhc2VVcmwgfSk7XG4gIH1cbn07Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6e30sImVudiI6e319fQ==), I just removed `: Promise<BrokenUrl[]>`, all `[...]`, and added one missing `}` to make it a valid JS code.

Can you find out which ESLint version are you using?

> We're using the @typescript-eslint/parser.
>
> as `page` is declared in a different scope than it is first assigned.

The rule checks if the assignment is in the same scope [here](https://github.com/eslint/eslint/blob/db1518374a5e88efedf1ed4609d879f3091af74f/lib/rules/prefer-const.js#L245), by using the scope analysis info, which is in this case provided by `@typescript-eslint/parser`. If you're using the latest versions of ESLint and `@typescript-eslint/parser`, this might be an issue for https://github.com/typescript-eslint/typescript-eslint.

(though, I also couldn't reproduce this with the latest `@typescript-eslint/parser`)

**Comment by strangedev** [comment]
Hey @mdjermanovic, thanks for checking!

I will try out your example and see if we maybe have a bug on our side.
